### PR TITLE
Add my groups overview

### LINKED ISSUE #426: My groups
Provide a "my groups" user route to list all the groups the user is a member.

### PR DESCRIPTION
**Pull request by MPParsley**
Fixes #426.

Here's what it would look like.

groups admin theme
groups frontend theme

### CODE REVIEW

File: src/OgMembershipListBuilder.php [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+
+namespace Drupal\og;
+
+use Drupal\Core\Entity\EntityInterface;
+use Drupal\Core\Entity\EntityListBuilder;
+use Drupal\Core\Entity\EntityTypeInterface;
+use Drupal\Core\Link;
+use Drupal\Core\Session\AccountProxyInterface;
+use Drupal\Core\Url;
+use Drupal\og\OgRoleInterface;
```
Comment (MPParsley) [inline src/OgMembershipListBuilder.php]:
```suggestion
```

File: src/OgMembershipListBuilder.php [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+    return $operations;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function buildRow(EntityInterface $entity) {
+    /** @var \Drupal\og\OgMembershipInterface $entity */
+    $row['label'] = Link::fromTextAndUrl($entity->getGroup()->label(), $entity->getGroup()->toUrl());
+    $row['roles'] = implode(', ', array_map(function(OgRoleInterface $role) {return $role->getLabel();}, $entity->getRoles()));
```
Comment (MPParsley) [inline src/OgMembershipListBuilder.php]:
```suggestion
    $callback = function(OgRoleInterface $role) {
      return $role->getLabel();
    }
    $row['roles'] = implode(', ', array_map($callback, $entity->getRoles()));
```

File: src/OgMembershipListBuilder.php [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+
+  /**
+   * {@inheritdoc}
+   */
+  public function buildRow(EntityInterface $entity) {
+    /** @var \Drupal\og\OgMembershipInterface $entity */
+    $row['label'] = Link::fromTextAndUrl($entity->getGroup()->label(), $entity->getGroup()->toUrl());
+    $callback = function(OgRoleInterface $role) {
+      return $role->getLabel();
+    }
```
Comment (MPParsley) [inline src/OgMembershipListBuilder.php]:
```suggestion
    };
```

File: src/OgMembershipListBuilder.php [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+    return $operations;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function buildRow(EntityInterface $entity) {
+    /** @var \Drupal\og\OgMembershipInterface $entity */
+    $row['label'] = Link::fromTextAndUrl($entity->getGroup()->label(), $entity->getGroup()->toUrl());
+    $callback = function(OgRoleInterface $role) {
```
Comment (MPParsley) [inline src/OgMembershipListBuilder.php]:
```suggestion
    $callback = function (OgRoleInterface $role) {
```

File: src/OgMembershipListBuilder.php [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+      return $role->getLabel();
+    };
+    $row['roles'] = implode(', ', array_map($callback, $entity->getRoles()));
+    $row['state'] = $entity->getState();
+
+    return $row + parent::buildRow($entity);
+  }
+
+  public function render() {
```
Comment (MPParsley) [inline src/OgMembershipListBuilder.php:92]:
```suggestion
  /**
   * {@inheritdoc}
   */
  public function render() {
```

File: src/OgMembershipListBuilder.php [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+    $build['table']['#empty'] = $this->t('There are no memberships yet.');
+    uasort($build['table']['#rows'], [$this, 'sort']);
+
+    return $build;
+  }
+
+  /**
+   * Sort results by group name.
+   *
+   * @param array $a
```
Comment (MPParsley) [inline src/OgMembershipListBuilder.php:103]:
```suggestion
   * @param array $a
   *   Group a.
```

File: src/OgMembershipListBuilder.php [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+   * @param array $b
```
Comment (MPParsley) [inline src/OgMembershipListBuilder.php:105]:
```suggestion
   * @param array $b
   *   Group b.
```

File: src/OgMembershipListBuilder.php [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+   *
+   * @return int
```
Comment (MPParsley) [inline src/OgMembershipListBuilder.php:108]:
```suggestion
   * @return int
   *   Returns the string comparison between the two labels of the groups.
```

File: tests/src/Kernel/Entity/OgStandardReferenceItemTest.php [hunk 9]
```diff
@@ -20,8 +20,25 @@ class OgStandardReferenceItemTest extends KernelTestBase {
...
+   * A list of bundles.
+   *
+   * @var string[]
+   */
   protected $bundles;
+
+  /**
+   * The machine name for a field.
+   *
+   * @var string[]
```
Comment (MPParsley) [inline tests/src/Kernel/Entity/OgStandardReferenceItemTest.php]:
```suggestion
   * @var string
```

File: og.routing.yml [hunk 10]
```diff
@@ -43,3 +43,17 @@ og.remove_multiple_roles_confirm:
...
+og.user_memberships:
+  path: '/user/{user}/groups'
+  defaults:
+    _entity_list: 'og_membership'
+    _title: 'My groups'
+  requirements:
+    user: '\d+'
+    _entity_access: 'user.view'
+  options:
+    _admin_route: 'TRUE'
```
Comment (matthijs-va) [inline og.routing.yml:56]: This isn't an admin route.

**Review by matthijs-va — CHANGES_REQUESTED** [review]

File: og.routing.yml (hunk 10 above)
Comment (MPParsley) [reply og.routing.yml:56]:
```suggestion
```

**Pull request merged**